### PR TITLE
Initial support for MSC3931: Room version push rule feature flags

### PR DESCRIPTION
**Pull request by turt2live**
## Notes for reviewer
**This is reviewable commit-by-commit (recommended).**

**The code in this PR is dead code - see other PRs (linked below) for full context.**

This PR's experimental flag does *not* match the implemented MSC's number - this is intentional, as the flag is used in future PRs.

## PR series

This PR is part of a series:
* https://github.com/matrix-org/synapse/pull/14520 - You are here.
* https://github.com/matrix-org/synapse/pull/14521
* https://github.com/matrix-org/synapse/pull/14524

### CODE REVIEW

File: rust/src/push/mod.rs [hunk 1]
```diff
@@ -491,6 +495,18 @@ fn test_deserialize_unstable_msc3664_condition() {
...
 
+#[test]
+fn test_deserialize_unstable_msc3931_condition() {
+    let json =
+        r#"{"kind":"org.matrix.msc3931.room_version_supports","feature":"org.example.feature"}"#;
+
+    let condition: Condition = serde_json::from_str(json).unwrap();
+    assert!(matches!(
+        condition,
+        Condition::Known(KnownCondition::RoomVersionSupports { feature: _ })
```
Comment (MatMaul) [inline rust/src/push/mod.rs:506]: Any reason to not try to match feature value `org.example.feature` too here ?

**Review by MatMaul — APPROVED** [review]
Minor remark on a test, otherwise LGTM.

File: rust/src/push/mod.rs (hunk 1 above)
Comment (turt2live) [reply rust/src/push/mod.rs:506]: "Rust yelled at me" 

If there's a way to do that, happy to do so :)

Comment (MatMaul) [reply rust/src/push/mod.rs:506]: So I tried `Cow::Borrowed("org.example.feature")` and the compiler yelling stopped, but the outcome was a failure, so I guess there are Cow subtleties I don't get.
Let's move on it's good enough like that.

**Review by MatMaul — APPROVED** [review]

**Pull request merged**